intermediate_source/pinmem_nonblock.py

```diff
@@ -0,0 +1,625 @@
+# pin memory manually from python main thread is a blocking operation on the host, and hence will annihilate much of
+# the benefit of using ``non_blocking=True`` (as CUDA does the `pin_memory` transfer anyway).
+#
+# One might now legitimately ask what use there is for the :meth:`~torch.Tensor.pin_memory` method.
+# In the following section, we will explore further how this can be used to accelerate the data transfer even more.
+#
+# Additional considerations
+# -------------------------
+#
+#   .. _pinmem_considerations:
+#
+# PyTorch notoriously provides a :class:`~torch.utils.data.DataLoader` class which constructor accepts a
+# ``pin_memory`` argument.
+# Considering our previous discussion on ``pin_memory``, you might wonder how the ``DataLoader`` manages to
+# accelerate data transfers if memory pinning is inherently blocking.
+#
+# The key lies in the DataLoader's use of a separate thread to handle the transfer of data from pageable to pinned
+# memory, thus preventing any blockage in the main thread.
+#
+# To illustrate this, we will use the TensorDict primitive from the homonymous library.
+# When invoking :meth:`~tensordict.TensorDict.to`, the default behavior is to send tensors to the device asynchronously,
+# followed by a single call to ``torch.device.synchronize()`` afterwards.
+#
+# Additionally, ``TensorDict.to()`` includes a ``non_blocking_pin`` option  which initiates multiple threads to execute
+# ``pin_memory()`` before proceeding with to ``to(device)``.
+# This approach can further accelerate data transfers, as demonstrated in the following example.
+#
+# .. code-block:: bash
+#
+#    # Install tensordict with the following command
+#    !pip3 install https://github.com/pytorch/tensordict
+#
+
+from tensordict import TensorDict
+import torch
+from torch.utils.benchmark import Timer
+
+# Create the dataset
+td = TensorDict({str(i): torch.randn(1_000_000) for i in range(1000)})
+
+# Runtimes
+copy_blocking = timer("td.to('cuda:0', non_blocking=False)")
+copy_non_blocking = timer("td.to('cuda:0')")
+copy_pin_nb = timer("td.to('cuda:0', non_blocking_pin=True, num_threads=0)")
+copy_pin_multithread_nb = timer("td.to('cuda:0', non_blocking_pin=True, num_threads=4)")
+
+# Rations
+r1 = copy_non_blocking / copy_blocking
+r2 = copy_pin_nb / copy_blocking
+r3 = copy_pin_multithread_nb / copy_blocking
+
+# Figure
+fig, ax = plt.subplots()
+
+xlabels = [0, 1, 2, 3]
+bar_labels = [
+    "Blocking copy (1x)",
+    f"Non-blocking copy ({r1:4.2f}x)",
+    f"Blocking pin, non-blocking copy ({r2:4.2f}x)",
+    f"Non-blocking pin, non-blocking copy ({r3:4.2f}x)",
+]
+values = [copy_blocking, copy_non_blocking, copy_pin_nb, copy_pin_multithread_nb]
+colors = ["tab:blue", "tab:red", "tab:orange", "tab:green"]
+
+ax.bar(xlabels, values, label=bar_labels, color=colors)
+
+ax.set_ylabel("Runtime (ms)")
+ax.set_title("Device casting runtime")
+ax.set_xticks([])
+ax.legend()
+
+plt.show()
```

Comment:
from this graph, it _still_ looks like pinning memory is useless because the red bar (the nonblock copy) is just as fast! so why would anyone ever use pin_memory still?

Reply:
that's a problem with the CI I think, we have a limited number of cores there.
On depgpu it renders better. That being said I did not manage to find anything where pin_memory() brought a real non-trivial benefit on any machine...

Reply:
Plus you can get some limited speed-up with limited threads on some machines, but at the cost of using more workers so your preproc/logging/whatever it is you use your CPUs for may become slower because of that. My take is that in general people should have a clear use case for pin-memory or not use it at all.

For instance, here's a scenario where I could consider pin-mem:
In distributed settings, you will sometimes pass a ton of tensors from worker 0 to worker 1, and worker 1 will send these tensors to CUDA.
With tensordict you could:
(1) consolidate all tensors in a single storage on worker 0
(2) send that single storage to worker 1 (faster than sending the tensors separately)
(3) receive that storage in pin-memory (not sure how to do that exactly, you'd need to hack the reductions methods appropriately, but I guess it's doable) - NOT receive in pageable mem and send to page-locked!
(4) send that to cuda

That will be much faster than sending all of the tensors, receiving them all and sending them to cuda with/without pin-mem.

@dstaay-fb

Reply:
(I wanted to call this tuto: don't use pin_memory but that was going a bit too far haha)